File: 21.재귀/2447.별 찍기 - 10/main.py
import sys

input = sys.stdin.readline


def recur(n, r_start, c_start):
    if n == 1:
        arr[r_start][c_start] = "*"
        return

    child_n = n // 3
    for r_offset in range(0, 3):
        for c_offset in range(0, 3):
            if r_offset == 1 and c_offset == 1:
                continue
            recur(child_n, r_start + child_n * r_offset, c_start + child_n * c_offset)


if __name__ == "__main__":
    N = int(input())

    arr = [[" " for i in range(N + 1)] for j in range(N + 1)]

    recur(N, 1, 1)

    for r in range(1, N + 1):
        print("".join(arr[r][1:]))

File: 21.재귀/2447.별 찍기 - 10/Main.java
import java.util.Scanner;

public class Main {

    static Character[][] arr;

    static void recur(int n, int rStart, int cStart) {
        if (n == 1) {
            arr[rStart][cStart] = '*';
            return;
        }

        int childN = n / 3;
        for (int rOffset = 0; rOffset < 3; rOffset++) {
            for (int cOffset = 0; cOffset < 3; cOffset++) {
                if (rOffset == 1 && cOffset == 1)
                    continue;

                recur(childN, rStart + childN * rOffset, cStart + childN * cOffset);
            }
        }
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int N = Integer.parseInt(sc.nextLine());

        arr = new Character[N + 1][N + 1];
        recur(N, 1, 1);

        StringBuilder sb = new StringBuilder();
        for (int i = 1; i < N + 1; i++) {
            for (int j = 1; j < N + 1; j++) {
                if (arr[i][j] == null) {
                    sb.append(" ");
                } else {
                    sb.append(arr[i][j]);
                }
            }
            sb.append("\n");
        }

        System.out.println(sb.toString());
        sc.close();
    }
}
